### DISCUSSION TITLE
Why use NamedTuple data type?

### DISCUSSION THREAD

**Discussion by HiramTheHero** [JuliaData/Tables.jl#268] [Q&A] [answered]
Hello All.

I hope you are well. This is a design question of Tables.jl. To my understanding, Tables.jl uses the NamedTuple data structure as its core type to work on. According to the help section on NamedTuples, it states "NamedTuples are immutable; neither the names nor the values can be modified in place after construction."

Because NamedTuple is immutable, I would think that this would cause some performance issues. (Such as, recreating a new NamedTuple anytime a change is performed.) What are the reasons why Tables.jl is built around NamedTuples instead of another, mutable, data structure?

I'm slowly gaining proficiency in Julia, so if my question is excessively naive, I apologize. Thank you for your time.

**Comment by bkamins** [comment]
`DataFrame` from DataFrames.jl is a mutable type you are asking for. The downside of being mutable is that the Julia compiler cannot easily optimize code using data stored in `DataFrame` (see https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations for more explanations).

**Comment by quinnj** [comment, marked as answer]
No worry about asking at all. A vector of NamedTuples or NamedTuple of vectors was chosen for two main reasons: 1) they're "builtin" to Julia, which means no additional dependency needed for Tables.jl, which aims to be a very _foundational_ package and 2) they're meant to be _simple_ table types to aid in testing or small, specific use-cases where they make sense. So definitely not meant as any kind of "full" table type or DataFrame, but also simple enough that they can be useful in certain workflows.

Hope that helps!

**Reply by HiramTheHero** [reply]
Thank you for the comprehensive answer. I guess I was misunderstanding the place of Tables.jl within the Julia Data Ecosystem. I was under the impression that Tables.jl was a "fundamental" datatype for packages that used it. So, all packages that used it had to inherit\use its structure. I didn't realize it was something separate. (I must have read the docs incorrectly, or something like that.) Thanks again for the answer. It is appreciated.

**Reply by quinnj** [reply]
No problem. Yeah, the primary role of Tables.jl is the _interfaces_ it defines, not the types it provides.
